sycl/unittests/program_manager/DynamicLinking.cpp

```diff
@@ -0,0 +1,192 @@
+#include <sycl/sycl.hpp>
+
+#include <helpers/MockKernelInfo.hpp>
+#include <helpers/PiImage.hpp>
+#include <helpers/PiMock.hpp>
+#include <helpers/RuntimeLinkingCommon.hpp>
+
+#include <gtest/gtest.h>
+
+namespace DynamicLinkingTest {
+class KernelA;
+class KernelB;
+class KernelC;
+class KernelD;
+class KernelE;
+} // namespace DynamicLinkingTest
+
+namespace sycl {
+inline namespace _V1 {
+namespace detail {
+#define KERNEL_INFO(KernelName)                                                \
+  template <>                                                                  \
+  struct KernelInfo<DynamicLinkingTest::KernelName>                            \
+      : public unittest::MockKernelInfoBase {                                  \
+    static constexpr const char *getName() { return #KernelName; }             \
+  };
+
+KERNEL_INFO(KernelA)
+KERNEL_INFO(KernelB)
+KERNEL_INFO(KernelC)
+KERNEL_INFO(KernelD)
+KERNEL_INFO(KernelE)
+
+#undef KERNEL_INFO
+} // namespace detail
+} // namespace _V1
+} // namespace sycl
+
+namespace {
+sycl::unittest::PiArray<sycl::unittest::PiProperty>
+createPropertySet(const std::vector<std::string> &Symbols) {
+  sycl::unittest::PiPropertySet PropSet;
+  sycl::unittest::PiArray<sycl::unittest::PiProperty> Props;
+  for (const std::string &Symbol : Symbols) {
+    std::vector<char> Storage(sizeof(pi_uint32));
+    uint32_t Val = 1;
+    auto *DataPtr = reinterpret_cast<char *>(&Val);
+    std::uninitialized_copy(DataPtr, DataPtr + sizeof(uint32_t),
+                            Storage.data());
+    sycl::unittest::PiProperty Prop(Symbol, Storage, PI_PROPERTY_TYPE_UINT32);
+
+    Props.push_back(Prop);
+  }
+  return Props;
+}
+
+sycl::unittest::PiImage
+generateImage(std::initializer_list<std::string> KernelNames,
+              const std::vector<std::string> &ExportedSymbols,
+              const std::vector<std::string> &ImportedSymbols,
+              unsigned char Magic,
+              sycl::detail::pi::PiDeviceBinaryType BinType =
+                  PI_DEVICE_BINARY_TYPE_SPIRV) {
+  sycl::unittest::PiPropertySet PropSet;
+  if (!ExportedSymbols.empty())
+    PropSet.insert(__SYCL_PI_PROPERTY_SET_SYCL_EXPORTED_SYMBOLS,
+                   std::move(createPropertySet(ExportedSymbols)));
+  if (!ImportedSymbols.empty())
+    PropSet.insert(__SYCL_PI_PROPERTY_SET_SYCL_IMPORTED_SYMBOLS,
+                   std::move(createPropertySet(ImportedSymbols)));
+  std::vector<unsigned char> Bin{Magic};
+
+  sycl::unittest::PiArray<sycl::unittest::PiOffloadEntry> Entries =
+      sycl::unittest::makeEmptyKernels(KernelNames);
+
+  sycl::unittest::PiImage Img{
+      BinType,
+      __SYCL_PI_DEVICE_BINARY_TARGET_SPIRV64, // DeviceTargetSpec
+      "",                                     // Compile options
+      "",                                     // Link options
+      std::move(Bin),
+      std::move(Entries),
+      std::move(PropSet)};
+
+  return Img;
+}
+
+static constexpr unsigned PROGRAM_A = 2;
```

Comment:
I'm not a huge fan of these names. We are testing  basic linking, unresolved dependencies, mutual dependencies, and ahead-of-time.  It'd be great if the names of these constants, kernels, and images conveyed these purposes throughout, rather than "Program B is for Kernel B which is for unresolved dependencies".